text/0000-sync-networked-vars.md

```diff
@@ -0,0 +1,110 @@
+
+It is proposed that the NetworkedVariable interpolation algorithm be easily customizable and that the user can chose whether it supports client or server side authoritative mechanics. To account for this, each NetworkedVaraible will keep track of the following:
+- When it was modified by a given client
+- When the server updated or transmitted it
+
+This will enable better client-side interpolation, lag compensation, and in the future provide the required flexibility for new systems (i.e. like snapshotting).
+
+This provides enough information such that the client can distinguish between its own updates from that of server-side adjustments (i.e. like physics).
+
+This will keep the overall bandwidth usage in check, preventing frequent updates for NetworkedVariables.
+
+This sets up MLAPI for upcoming features like snapshot.
+
+# Dependency
+[dependency]: #dependency
+
+This feature depends on a Tick System that would provide:
+- A discrete measurement of the local time, based on an integer number of tick elapsed
+- The tick number being updated at the very beginning of each frame and stable throughout the frame
+- A tick that is based on the real time elapsed since a fixed point in the past 
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Some definitions:
+
+#### Local tick
+The tick number at which a variable was modified, relative to the tick system on this machine. If I get an update for a variable I changed, this tell me: when did I change it?
+
+#### Remote tick
+The tick number at which a variable was relayed, relative to the tick system on the remote machine. If I get an update from the server, this tells me the local tick on the server when it sent the update.
+
+#### Network Tick wrapping 
+
+The tick value sent will be a rolling value, probably 8- or 16-bit. This will be hidden internally, but will keep the bandwidth usage in check.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+We suggest changing:
+
+- in `Runtime/NetworkedVar/INetworkedVar.cs`, `INetworkedVar` `ReadField()` to become
+
+      public void ReadField(Stream stream, ushort localTick, ushort remoteTick);
+
+- in `Runtime/NetworkedVar/INetworkedVar.cs`, `INetworkedVar` `ReadDelta()` to become
+
+      public void ReadDelta(Stream stream, bool keepDirtyDelta, ushort localTick, ushort remoteTick);
+
+- `NetworkedVar` to include two new properties:
+
+      public ushort LocalTick { get; internal set; }
+      public ushort RemoteTick { get; internal set; }
+      
+ - `SendTickrate` and `LastSyncedTime` to be removed from `NetworkedVar`
+
+Below diagram shows a couple NetworkedVariable updates with example tick numbers.
+
+![](0000-sync-networked-vars/netvars.png)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This increases slightly the bandwidth usage of _individual_ NetworkedVariable update it also renders the NetworkedVariables system dependent on the tick system.
```

Comment:
A more generalized/global ticking & snapshotting could help in this area greatly but we're not there yet...